File: Day-2/Labs/kafka-python-lab/simple_producer.py
"""
Publish sample messages to Kafka.

Run:
    python simple_producer.py
    python simple_producer.py localhost:9092 lab-messages 3
"""

import sys
import time

from kafka import KafkaProducer
from kafka.errors import KafkaError

from config import DEFAULT_BOOTSTRAP, DEFAULT_MESSAGE_COUNT, DEFAULT_TOPIC


def main() -> None:
    bootstrap = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_BOOTSTRAP
    topic = sys.argv[2] if len(sys.argv) > 2 else DEFAULT_TOPIC
    message_count = int(sys.argv[3]) if len(sys.argv) > 3 else DEFAULT_MESSAGE_COUNT

    print(f"Connecting to {bootstrap}, topic={topic}, messages={message_count}")

    producer = KafkaProducer(
        bootstrap_servers=bootstrap,
        value_serializer=lambda v: v.encode("utf-8"),
        client_id="python-lab-producer",
        acks="all",
    )

    try:
        for i in range(1, message_count + 1):
            value = f"Hello from Python producer - message {i}"
            future = producer.send(topic, value=value)

            try:
                metadata = future.get(timeout=10)
                print(
                    f'Sent: "{value}" -> '
                    f"partition={metadata.partition} offset={metadata.offset}"
                )
            except KafkaError as exc:
                print(f"Failed to send message {i}: {exc}", file=sys.stderr)
                sys.exit(1)

        producer.flush()
    finally:
        producer.close()

    print("Producer finished.")


if __name__ == "__main__":
    main()

File: Day-2/Labs/kafka-dotnet-lab/src/KafkaDotNetLab/SimpleProducer.cs
using Confluent.Kafka;

namespace KafkaDotNetLab;

internal static class SimpleProducer
{
    public static async Task RunAsync(string[] args)
    {
        var bootstrap = KafkaConfig.Arg(args, 0, KafkaConfig.DefaultBootstrap);
        var topic = KafkaConfig.Arg(args, 1, KafkaConfig.DefaultTopic);
        var messageCount = KafkaConfig.ParseInt(KafkaConfig.Arg(args, 2, "5"), 5);

        Console.WriteLine(
            $"Connecting to {bootstrap}, topic={topic}, messages={messageCount}");

        var config = new ProducerConfig
        {
            BootstrapServers = bootstrap,
            ClientId = "dotnet-lab-producer",
            Acks = Acks.All,
        };

        using var producer = new ProducerBuilder<string, string>(config).Build();

        for (var i = 1; i <= messageCount; i++)
        {
            var value = $"Hello from .NET producer - message {i}";

            try
            {
                var result = await producer.ProduceAsync(
                    topic,
                    new Message<string, string> { Value = value });

                Console.WriteLine(
                    $"Sent: \"{value}\" -> partition={result.Partition.Value} offset={result.Offset.Value}");
            }
            catch (ProduceException<string, string> ex)
            {
                Console.Error.WriteLine($"Failed to send message {i}: {ex.Error.Reason}");
                Environment.Exit(1);
            }
        }

        producer.Flush(TimeSpan.FromSeconds(10));
        Console.WriteLine("Producer finished.");
    }
}
